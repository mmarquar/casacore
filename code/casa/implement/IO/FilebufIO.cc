//# FilebufIO.cc: Class for IO on a file using a filebuf object.
//# Copyright (C) 1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/aips.h>
#include <aips/IO/FilebufIO.h>
#include <aips/Utilities/Assert.h>
#include <aips/Exceptions/Error.h>
#include <iostream.h>
#include <fstream.h>
#include <fcntl.h>

FilebufIO::FilebufIO()
: itsOwner     (False),
  itsFilebuf   (0),
  itsBuffer    (0),
  itsSeekable  (False),
  itsReadable  (False),
  itsWritable  (False),
  itsReadDone  (False),
  itsWriteDone (False)
{}

FilebufIO::FilebufIO (int fd, uInt bufferSize)
: itsFilebuf (0)
{
    attach (fd, bufferSize);
}

FilebufIO::~FilebufIO()
{
    detach();
}

void FilebufIO::prepareAttach (uInt bufferSize, Bool readable, Bool writable)
{
    AlwaysAssert (itsFilebuf == 0, AipsError);
    itsOwner     = True;
    itsReadable  = readable;
    itsWritable  = writable;
    itsReadDone  = True;
    itsWriteDone = True;
    // Create a buffer for filebuf and attach it.
    itsBuffer  = new char[bufferSize];
    itsFilebuf = new filebuf;
    AlwaysAssert (itsBuffer != 0  &&  itsFilebuf != 0, AipsError);
    AlwaysAssert (itsFilebuf->setbuf(itsBuffer, bufferSize) != 0, AipsError);
}

void FilebufIO::attach (int fd, uInt bufferSize)
{
    prepareAttach (bufferSize, False, False);
    // Attach the file.
    AlwaysAssert (itsFilebuf->attach (fd) != 0, AipsError);
    int flags = fcntl (fd, F_GETFL);
    if (flags & O_RDWR) {
	itsReadable = True;
	itsWritable = True;
    } else if (flags & O_WRONLY) {
	itsWritable = True;
    } else {
	itsReadable = True;
    }
    fillSeekable();
}

void FilebufIO::detach()
{
    if (itsFilebuf != 0) {
	if (itsOwner) {
	    delete itsFilebuf;
	    delete [] itsBuffer;
	}
	itsFilebuf = 0;
	itsBuffer  = 0;
    }
}

void FilebufIO::fillSeekable()
{
    itsSeekable = ToBool (seek (0, ByteIO::Current)  >= 0);
}


String FilebufIO::fileName() const
{
    return "";
}


void FilebufIO::write (uInt size, const void* buf)
{
    // Throw an exception if not writable.
    if (!itsWritable) {
	throw (AipsError ("FilebufIO object is not writable"));
    }
    // If a read has been done, adjust the write pointer.
    if (itsReadDone) {
	itsFilebuf->seekoff (0, ios::cur, ios::in|ios::out);
	itsReadDone = False;
    }
    if (itsFilebuf->sputn ((const char*)buf, size) != size) {
	throw (AipsError ("FilebufIO: error while writing " + fileName()));
    }
    itsWriteDone = True;
}

void FilebufIO::read (uInt size, void* buf)
{
    // Throw an exception if not readable.
    if (!itsReadable) {
	throw (AipsError ("FilebufIO object is not readable"));
    }
    // If a write has been done, adjust the read pointer.
    if (itsWriteDone) {
	itsFilebuf->seekoff (0, ios::cur, ios::in|ios::out);
	itsWriteDone = False;
    }
    if (itsFilebuf->sgetn ((char*)buf, size) != size) {
	throw (AipsError ("FilebufIO: error while reading " + fileName()));
    }
    itsReadDone = True;
}

Long FilebufIO::seek (Long offset, ByteIO::SeekOption dir)
{
    itsReadDone  = False;
    itsWriteDone = False;
    switch (dir) {
    case ByteIO::Begin:
	return itsFilebuf->seekoff (offset, ios::beg, ios::in|ios::out);
    case ByteIO::End:
	return itsFilebuf->seekoff (offset, ios::end, ios::in|ios::out);
    }
    return itsFilebuf->seekoff (offset, ios::cur, ios::in|ios::out);
}

Long FilebufIO::length()
{
    // Get current position to be able to reposition.
    Long pos = seek (0, ByteIO::Current);
    // Seek to the end of the stream.
    // If it fails, we cannot seek and the current position is the length.
    Long len = seek (0, ByteIO::End);
    if (len < 0) {
	return pos;
    }
    // Reposition and return the length.
    seek (pos, ByteIO::Begin);
    return len;
}

   
Bool FilebufIO::isReadable() const
{
    return itsReadable;
}

Bool FilebufIO::isWritable() const
{
    return itsWritable;
}

Bool FilebufIO::isSeekable() const
{
    return itsSeekable;
}
