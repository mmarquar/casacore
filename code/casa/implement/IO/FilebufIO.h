//# FilebufIO.h: Class for IO on a file using standard IO
//# Copyright (C) 1996,1997,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_FILEBUFIO_H)
#define AIPS_FILEBUFIO_H


//# Includes
#include <aips/aips.h>
#include <aips/IO/ByteIO.h>
#include <aips/Utilities/String.h>
#include <stdio.h>

// <summary> Class for IO on a file using standard IO.</summary>

// <use visibility=export>

// <reviewed reviewer="Friso Olnon" date="1996/11/06" tests="tByteIO" demos="">
// </reviewed>

// <prerequisite> 
//    <li> <linkto class=ByteIO>ByteIO</linkto> class
//    <li> stdio package (see man stdio)
// </prerequisite>

// <synopsis> 
// This class is a specialization of class
// <linkto class=ByteIO>ByteIO</linkto>. It uses <src>stdio</src>
// to read/write data.
// <p>
// <src>stdio</src> uses an internal buffer. The size of the
// buffer influences the performance. Usually it is best to use
// the default buffer size (which stdio optimizes for the disk).
// However, it is possible to specify a particular buffer size.
// <p>
// It is also possible to construct a <src>FilebufIO</src> object
// from a file descriptor (e.g. for a pipe or socket).
// The constructor will determine automatically if the file is
// readable, writable and seekable.
// </synopsis>

// <example>
// This example shows how FilebufIO can be used with an fd.
// It uses the fd for a regular file, which could be done in an easier
// way using class <linkto class=RegularFileIO>RegularFileIO</linkto>.
// However, when using pipes or sockets, this would be the only way.
// <srcblock>
//    // Get a file descriptor for the file.
//    int fd = open ("file.name");
//    // Use that as the source of AipsIO (which will also use CanonicalIO).
//    FilebufIO fio (fd);
//    AipsIO stream (&fio);
//    // Read the data.
//    Int vali;
//    Bool valb;
//    stream >> vali >> valb;
// </srcblock>
// </example>

// <motivation> 
// Make it possible to use the AIPS++ IO functionality on any file.
// In this way any device can be hooked to the IO framework.
// </motivation>


class FilebufIO: public ByteIO
{
public: 
    // Default constructor.
    // A stream can be attached using the attach function.
    FilebufIO();

    // Construct from a given file pointer which is taken over
    // (thus the file is closed by function detach or the destructor).
    // It will determine itself if the file is readable, writable,
    // and seekable.
    // A buffer with the given length will be used by the FILE object.
    // A zero buffer size means an appropriate buffer size will be used.
    explicit FilebufIO (FILE*, uInt bufferSize=0);

    // Construct from a given file pointer.
    // When takeOver is true, the file will be closed by the
    // detach function or the destructor.
    // It will determine itself if the file is readable, writable,
    // and seekable.
    FilebufIO (FILE*, Bool takeOver);

    // Construct from the given file descriptor.
    // Note that the destructor and the detach function implicitly close
    // the file descriptor.
    explicit FilebufIO (int fd, uInt bufferSize=16384);

    // Attach the given file pointer which is taken over
    // (thus the file is closed by function detach or the destructor).
    // It will determine itself if the file is readable, writable,
    // and seekable.
    // A buffer with the given length will be used by the FILE object.
    // A zero buffer size means an appropriate buffer size will be used.
    void attach (FILE*, uInt bufferSize=0);

    // Attach the given file pointer.
    // When takeOver is true, the file will be closed by the
    // detach function or the destructor.
    // It will determine itself if the file is readable, writable,
    // and seekable.
    void attach (FILE*, Bool takeOver);

    // Attach to the given file descriptor.
    // Note that the destructor and the detach function implicitly close
    // the file descriptor.
    void attach (int fd, uInt bufferSize=16384);

    // The destructor closes the file when it was owned and opened and not
    // closed yet.
    ~FilebufIO();
    
    // Write the number of bytes.
    virtual void write (uInt size, const void* buf);

    // Read <src>size</src> bytes from the File. Returns the number of bytes
    // actually read. Will throw an Exception (AipsError) if the requested
    // number of bytes could not be read unless throwException is set to
    // False. Will always throw an exception if the file is not readable or
    // the system call returns an undocumented value.
    virtual Int read (uInt size, void* buf, Bool throwException=True);    

    // Get the length of the byte stream.
    virtual Int64 length();
       
    // Is the IO stream readable?
    virtual Bool isReadable() const;

    // Is the IO stream writable?
    virtual Bool isWritable() const;

    // Is the IO stream seekable?
    virtual Bool isSeekable() const;

    // Get the file name of the file attached.
    virtual String fileName() const;


protected:
    // Attach the given FILE pointer.
    // If bufferSize>0, it allocates a buffer for the file.
    void attach (FILE* file, uInt bufferSize, Bool readable, Bool writable);

    // Detach the FILE. Close it when it is owned.
    void detach();

    // Determine if the file descriptor is readable and/or writable.
    void fillRWFlags (int fd);

    // Determine if the file is seekable.
    void fillSeekable();

    // Reset the position pointer to the given value. It returns the
    // new position.
    virtual Int64 doSeek (Int64 offset, ByteIO::SeekOption);

    // Get the file for the derived class.
    FILE* getFilePtr();

    // Get the buffer size used.
    uInt bufferSize() const;

private:
    Bool        itsOwner;
    Bool        itsSeekable;
    Bool        itsReadable;
    Bool        itsWritable;
    FILE*       itsFile;
    uInt        itsBufSize;
    char*       itsBuffer;
    Bool        itsReadDone;
    Bool        itsWriteDone;

    // Copy constructor, should not be used.
    FilebufIO (const FilebufIO& that);

    // Assignment, should not be used.
    FilebufIO& operator= (const FilebufIO& that);
};


inline FILE* FilebufIO::getFilePtr()
{
    return itsFile;
}
inline uInt FilebufIO::bufferSize() const
{
    return itsBufSize;
}


#endif
