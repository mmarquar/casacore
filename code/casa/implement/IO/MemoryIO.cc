//# MemoryIO.cc: Class for IO in memory
//# Copyright (C) 1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <aips/IO/MemoryIO.h>
#include <aips/Utilities/Assert.h>
#include <aips/Exceptions/Error.h>


MemoryIO::MemoryIO (uLong initialSize, uLong expandSize)
: itsBuffer     (0),
  itsAlloc      (initialSize),
  itsExpandSize (expandSize),
  itsUsed       (0),
  itsPosition   (0),
  itsReadable   (True),
  itsWritable   (True),
  itsCanDelete  (True)
{
    if (itsAlloc > 0) {
	itsBuffer = new uChar[itsAlloc];
	AlwaysAssert (itsBuffer != 0, AipsError);
    }
}

MemoryIO::MemoryIO (const void* buffer, uLong size)
: itsBuffer     ((uChar*)buffer),
  itsAlloc      (size),
  itsExpandSize (0),
  itsPosition   (0),
  itsUsed       (size),
  itsReadable   (True),
  itsWritable   (False),
  itsCanDelete  (False)
{}

MemoryIO::MemoryIO (void* buffer, uLong size, ByteIO::OpenOption option,
		    uLong expandSize, Bool canDelete)
: itsBuffer     ((uChar*)buffer),
  itsAlloc      (size),
  itsExpandSize (expandSize),
  itsPosition   (0),
  itsUsed       (size),
  itsReadable   (True),
  itsWritable   (True),
  itsCanDelete  (canDelete)
{
    // Make sure there is a buffer.
    if (itsAlloc > 0) {
	AlwaysAssert (itsBuffer != 0, AipsError);
    }
    // Adapt position, etc. from the option.
    switch (option) {
    case ByteIO::Old:
	itsWritable = False;
	break;
    case ByteIO::Append:
	itsPosition = itsUsed;
	break;
    case ByteIO::New:
    case ByteIO::NewNoReplace:
    case ByteIO::Scratch:
	itsUsed = 0;
    }
}


MemoryIO::~MemoryIO()
{
    if (itsCanDelete) {
	delete [] itsBuffer;
    }
}

void MemoryIO::write (uInt size, const void* buf)
{
    // Throw an exception if not writable.
    if (!itsWritable) {
	throw (AipsError ("MemoryIO object is not writable"));
    }
    // Expand the buffer when needed (and possible).
    uLong minSize = itsPosition + size;
    if (minSize > itsAlloc) {
	if (! expand (minSize)) {
	    throw (AipsError ("MemoryIO::write: buffer cannot be expanded"));
	}
    }
    // Copy the data and set new position and used.
    memcpy (itsBuffer + itsPosition, buf, size);
    itsPosition += size;
    if (itsPosition > itsUsed) {
	itsUsed = itsPosition;
    }
}

void MemoryIO::read (uInt size, void* buf)
{
    // Throw an exception if not readable.
    if (!itsReadable) {
	throw (AipsError ("MemoryIO object is not readable"));
    }
    if (itsPosition + size > itsUsed) {
	throw (AipsError ("MemoryIO::read beyond end-of-buffer"));
    }
    memcpy (buf, itsBuffer + itsPosition, size);
    itsPosition += size;
}

Long MemoryIO::seek (Long offset, ByteIO::SeekOption dir)
{
    // Determine the new position.
    // Exit with error status if negative.
    Long newPos;
    switch (dir) {
    case ByteIO::Begin:
	newPos = offset;
	break;
    case ByteIO::End:
	newPos = itsUsed + offset;
	break;
    default:
	newPos = itsPosition + offset;
	break;
    }
    if (newPos < 0) {
	throw (AipsError ("MemoryIO::seek before start of object"));
    }
    // It is possible to seek past the end of the buffer.
    // This means that the buffer usage increases.
    // Expand the buffer if needed.
    // Initialize the new buffer positions with zeroes.
    if (newPos > itsUsed) {
	// Throw an exception if not writable.
	if (!itsWritable) {
	    throw (AipsError ("MemoryIO::seek past end of readonly object"));
	}
	if (newPos > itsAlloc) {
	    if (! expand (newPos)) {
		throw (AipsError("MemoryIO::seek: buffer cannot be expanded"));
	    }
	}
	for (; itsUsed<newPos; itsUsed++) {
	    itsBuffer[itsUsed] = 0;
	}
    }
    itsPosition = newPos;
    return newPos;
}


Bool MemoryIO::expand (uLong minSize)
{
    // Check if expansion is really needed.
    if (minSize <= itsAlloc) {
	return True;
    }
    // Return with error status when expansion is not possible.
    if (itsExpandSize == 0) {
	return False;
    }
    // Expand with at least the expansion size.
    if (minSize < itsAlloc + itsExpandSize) {
	minSize = itsAlloc + itsExpandSize;
    }
    // Allocate new buffer, copy contents and delete old buffer (if possible).
    uChar* newBuffer = new uChar[minSize];
    AlwaysAssert (newBuffer != 0, AipsError);
    // Copy the old contents (if any).
    if (itsBuffer != 0) {
	memcpy (newBuffer, itsBuffer, itsUsed);
	if (itsCanDelete) {
	    delete [] itsBuffer;
	}
    }
    itsBuffer    = newBuffer;
    itsAlloc     = minSize;
    itsCanDelete = True;
    return True;
}


Long MemoryIO::length()
{
    return itsUsed;
}

Bool MemoryIO::isReadable() const
{
    return itsReadable;
}
Bool MemoryIO::isWritable() const
{
    return itsWritable;
}
Bool MemoryIO::isSeekable() const
{
    return True;
}
