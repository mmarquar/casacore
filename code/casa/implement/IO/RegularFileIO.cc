//# RegularFileIO.cc: Class for IO on a regular file
//# Copyright (C) 1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/aips.h>
#include <aips/IO/RegularFileIO.h>
#include <aips/Utilities/String.h>
#include <aips/Utilities/Assert.h>
#include <aips/Exceptions/Error.h>
#include <unistd.h>
#include <fcntl.h>


RegularFileIO::RegularFileIO (const RegularFile& regularFile,
			      ByteIO::OpenOption option, uInt bufferSize)
: itsOption      (option),
  itsRegularFile (regularFile)
{
    // Open file as input and/or output.
    Bool writable = True;
    int stropt;
    switch (option) {
    case ByteIO::Old:
	writable = False;
	stropt = ios::in|ios::nocreate;
	break;
    case ByteIO::New:
    case ByteIO::Scratch:
	stropt = ios::in|ios::out|ios::trunc;
	break;
    case ByteIO::NewNoReplace:
	stropt = ios::in|ios::out|ios::noreplace|ios::trunc;
	break;
    case ByteIO::Append:
	stropt = ios::in|ios::out|ios::app;
	break;
    case ByteIO::Update:
    case ByteIO::Delete:
	stropt = ios::in|ios::out|ios::nocreate;
	break;
    default:
	throw (AipsError ("RegularFileIO: unknown open option"));
    }
    prepareAttach (bufferSize, True, writable);
    // Open the file.
    const String& name = itsRegularFile.path().expandedName();
    if (getFilebuf()->open (name.chars(), stropt) == 0) {
	throw (AipsError ("RegularFileIO: error in open or create of file " +
			  name));
    }
    fillSeekable();
}

RegularFileIO::~RegularFileIO()
{
    getFilebuf()->close();
    detach();
    if (itsOption == ByteIO::Scratch  ||  itsOption == ByteIO::Delete) {
	itsRegularFile.remove();
    }
}


void RegularFileIO::reopenRW()
{
    if (isWritable()) {
	return;
    }
    // First try if the file can be opened as read/write.
    const String& name = itsRegularFile.path().expandedName();
    int fd = ::open (name.chars(), O_RDWR);
    if (fd == -1) {
	throw (AipsError ("RegularFileIO: reopenRW not possible for file " +
			  name));
    }
    ::close (fd);
    // It can be reopened, so close and reopen.
    getFilebuf()->close();
    AlwaysAssert (getFilebuf()->open (name.chars(),
				      ios::in|ios::out|ios::nocreate) != 0,
		  AipsError);
    setWritable();
    itsOption = ByteIO::Update;
}


String RegularFileIO::fileName() const
{
    return itsRegularFile.path().expandedName();
}
