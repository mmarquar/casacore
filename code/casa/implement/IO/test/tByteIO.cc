//# tByteIO.cc: Test program for class ByteIO and derived classes
//# Copyright (C) 1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/aips.h>
#include <aips/IO/FilebufIO.h>
#include <aips/IO/RegularFileIO.h>
#include <aips/IO/MemoryIO.h>
#include <aips/OS/RegularFile.h>
#include <aips/Utilities/Assert.h>
#include <aips/Exceptions/Error.h>
#include <fcntl.h>


void checkLength (ByteIO& fio, uInt& curLength, uInt addLength)
{
    curLength += addLength;
    AlwaysAssertExit (fio.length() == curLength);
}


static Bool valb=True;
static Short vals=-3;
static uShort valus=2;
static Int vali=1000;
static uInt valui=32768;
static Long vall=-14793;
static uLong valul=17;
static float valf=1.2;
static double vald=-3.14;

void checkValues (ByteIO& fio, uShort incr)
{
    fio.seek (0);
    uInt curLength = fio.length();
    Bool resb;
    fio.read (sizeof(Bool), &resb);
    AlwaysAssertExit (resb == valb);
    Short ress;
    fio.read (sizeof(Short), &ress);
    AlwaysAssertExit (ress == vals - incr);
    uShort resus;
    fio.read (sizeof(uShort), &resus);
    AlwaysAssertExit (resus == valus + incr);
    Int resi;
    fio.read (sizeof(Int), &resi);
    AlwaysAssertExit (resi == vali);
    uInt resui;
    fio.read (sizeof(uInt), &resui);
    AlwaysAssertExit (resui == valui);
    Long resl;
    fio.read (sizeof(Long), &resl);
    AlwaysAssertExit (resl == vall);
    uLong resul;
    fio.read (sizeof(uLong), &resul);
    AlwaysAssertExit (resul == valul);
    float resf;
    fio.read (sizeof(float), &resf);
    AlwaysAssertExit (resf == valf);
    double resd;
    fio.read (sizeof(double), &resd);
    AlwaysAssertExit (resd == vald);
    AlwaysAssertExit (fio.length() == curLength);
}

void doIt (ByteIO& fio)
{
    uInt length = 0;
    AlwaysAssertExit (fio.length() == 0);
    fio.write (sizeof(Bool), &valb);
    checkLength (fio, length, sizeof(Bool));
    fio.write (sizeof(Short), &vals);
    checkLength (fio, length, sizeof(Short));
    fio.write (sizeof(uShort), &valus);
    checkLength (fio, length, sizeof(uShort));
    fio.write (sizeof(Int), &vali);
    checkLength (fio, length, sizeof(Int));
    fio.write (sizeof(uInt), &valui);
    checkLength (fio, length, sizeof(uInt));
    fio.write (sizeof(Long), &vall);
    checkLength (fio, length, sizeof(Long));
    fio.write (sizeof(uLong), &valul);
    checkLength (fio, length, sizeof(uLong));
    fio.write (sizeof(float), &valf);
    checkLength (fio, length, sizeof(float));
    fio.write (sizeof(double), &vald);
    checkLength (fio, length, sizeof(double));
    
    checkValues (fio, 0);
    
    fio.seek (sizeof(Bool));
    AlwaysAssertExit (fio.length() == length);
    uShort incr = 100;
    Short vals1 = vals - incr;
    Short ress;
    fio.write (sizeof(Short), &vals1);
    fio.seek (sizeof(Bool));
    fio.read (sizeof(Short), &ress);
    AlwaysAssertExit (ress == vals1);
    uShort valus1 = valus + incr;
    uShort resus;
    fio.write (sizeof(uShort), &valus1);
    fio.seek (-sizeof(uShort), ByteIO::Current);
    fio.read (sizeof(uShort), &resus);
    AlwaysAssertExit (resus == valus1);
    Int resi;
    fio.read (sizeof(Int), &resi);
    AlwaysAssertExit (resi == vali);
    AlwaysAssertExit (fio.length() == length);
    
    checkValues (fio, incr);
}

void checkReopen()
{
    RegularFile rfile("tByteIO_tmp.data");
    {
	RegularFileIO fio(rfile);
	checkValues (fio, 100);
	fio.reopenRW();
	fio.seek (sizeof(Bool));
	Short vals;
	fio.read (sizeof(Short), &vals);
	vals -= 50;
	fio.seek (sizeof(Bool));
	fio.write (sizeof(Short), &vals);
	uShort valus;
	fio.read (sizeof(uShort), &valus);
	valus += 50;
	fio.seek (-sizeof(uShort), ByteIO::Current);
	fio.write (sizeof(uShort), &valus);
	checkValues (fio, 150);
    }

    rfile.setPermissions (0444);
    RegularFileIO fio2(rfile);
    checkValues (fio2, 150);
    Bool flag = False;
    try {
	fio2.reopenRW();
    } catch (AipsError x) {
	flag = True;
    } end_try;
    AlwaysAssertExit (flag);
    checkValues (fio2, 150);
    rfile.setPermissions (0644);
}

void testMemoryIO()
{
    {
	uChar buf[10];
	MemoryIO membuf (buf, sizeof(buf), ByteIO::New, 6);
	doIt (membuf);
	AlwaysAssertExit (membuf.getBuffer() != (const uChar*)&buf);
	uLong length = membuf.length();
	uInt incr = 20;
	membuf.seek (incr, ByteIO::End);
	AlwaysAssertExit (membuf.length() == length+incr);
	checkValues (membuf, 100);
	char val;
	Long lincr = incr;
	membuf.seek (-lincr, ByteIO::End);
	for (int i=0; i<incr; i++) {
	    membuf.read (1, &val);
	    AlwaysAssertExit (val == 0);
	}
	try {
	    membuf.read (1, &val);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;         // read beyond object
	} end_try;
	try {
	    membuf.seek (-(length + incr + 1), ByteIO::Current);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;         // negative seek
	} end_try;
	membuf.seek (-(length + incr), ByteIO::Current);
    }
    {
	char buf[10];
	MemoryIO membuf (buf, sizeof(buf), ByteIO::New, 0);
	try {
	    doIt (membuf);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} end_try;
	try {
	    membuf.seek (10, ByteIO::End);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} end_try;
	AlwaysAssertExit (membuf.getBuffer() == (const uChar*)buf);
    }
    {
	char* buf = new char[10];
	MemoryIO membuf (buf, 10, ByteIO::Scratch, 0, True);
	try {
	    doIt (membuf);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} end_try;
	try {
	    membuf.seek (10, ByteIO::End);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} end_try;
	AlwaysAssertExit (membuf.getBuffer() == (const uChar*)buf);
    }
}


main()
{
    try {
	testMemoryIO();

	MemoryIO file2;
	doIt (file2);
	
	MemoryIO file3 (file2.getBuffer(), file2.length());
	try {
	    file3.write (0, 0);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;          // readonly
	} end_try;
	checkValues (file3, 100);
	
	{
	    RegularFileIO file1(RegularFile("tByteIO_tmp.data"), ByteIO::New);
	    doIt (file1);
	}
	checkReopen();

	int fd = open ("tByteIO_tmp.data2", O_CREAT|O_TRUNC|O_RDWR, 0644);
	int flags = fcntl (fd, F_GETFL);
	if (flags & O_RDWR) {
	    cout << "read/write" << endl;
	} else if (flags & O_WRONLY) {
	    cout << "writeonly" << endl;
	} else {
	    cout << "readonly" << endl;
	}
	FilebufIO file4 (fd, ByteIO::New);
	doIt (file4);

	int fd1 = open ("tByteIO_tmp.data2", O_RDONLY, 0644);
	int flags1 = fcntl (fd1, F_GETFL);
	if (flags1 & O_RDWR) {
	    cout << "read/write" << endl;
	} else if (flags1 & O_WRONLY) {
	    cout << "writeonly" << endl;
	} else {
	    cout << "readonly" << endl;
	}
	FilebufIO file5 (fd1);
	checkValues (file5, 100);

	int fd2 = creat ("tByteIO_tmp.data2", 0644);
	int flags2 = fcntl (fd2, F_GETFL);
	if (flags2 & O_RDWR) {
	    cout << "read/write" << endl;
	} else if (flags2 & O_WRONLY) {
	    cout << "writeonly" << endl;
	} else {
	    cout << "readonly" << endl;
	}
	FilebufIO file6 (fd2);
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } end_try;
    return 0;                           // exit with success status
}

