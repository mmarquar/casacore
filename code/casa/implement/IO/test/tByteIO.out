MemoryIO::read beyond end-of-buffer
MemoryIO::seek before start of object
MemoryIO::write: buffer cannot be expanded
MemoryIO::seek: buffer cannot be expanded
MemoryIO::write: buffer cannot be expanded
MemoryIO::seek: buffer cannot be expanded
MemoryIO object is not writable
read/write
readonly
writeonly
